src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -499,6 +509,118 @@ public override string MetadataName
         /// </summary>
         internal abstract bool HasDeclaredRequiredMembers { get; }
 
+#nullable enable
+        /// <summary>
+        /// Whether the type encountered an error while trying to build its complete list of required members.
+        /// </summary>
+        internal bool HasRequiredMembersError
+        {
+            get
+            {
+                CalculateRequiredMembersIfRequired();
+                Debug.Assert(!_lazyRequiredMembers.IsDefault);
+                return _lazyHasRequiredMembersErrorDoNotAccessDirectly;
+            }
+        }
+
+        /// <summary>
+        /// The full list of all required members for this type, including from base classes. If <see cref="HasRequiredMembersError"/> is true,
+        /// this returns empty.
+        /// </summary>
+        /// <remarks>
+        /// Do not call this API if all you need are the required members declared on this type. Use <see cref="GetMembers()"/> instead, filtering for
+        /// required members, instead of calling this API.
+        /// </remarks>
+        internal ImmutableSegmentedDictionary<string, Symbol> AllRequiredMembers
+        {
+            get
+            {
+                CalculateRequiredMembersIfRequired();
+                Debug.Assert(!_lazyRequiredMembers.IsDefault);
+                return _lazyRequiredMembers;
+            }
+        }
+
+        private void CalculateRequiredMembersIfRequired()
+        {
+            if (_lazyRequiredMembers.IsDefault)
+            {
+                ImmutableSegmentedDictionary<string, Symbol>.Builder? builder = null;
+                bool success = TryCalculateRequiredMembers(ref builder);
+                var requiredMembers = success
+                    ? builder?.ToImmutable() ?? ImmutableSegmentedDictionary<string, Symbol>.Empty
+                    : ImmutableSegmentedDictionary<string, Symbol>.Empty;
+
+                _lazyHasRequiredMembersErrorDoNotAccessDirectly = !success;
+                // InterlockedInitialize uses `Interlocked.CompareExchange` under the hood, which will introduce a full
+                // memory barrier, ensuring that the _lazyHasRequiredMembersErrorDoNotAccessDirectly write is not reordered
+                // after the initialization of _lazyRequiredMembers.
+                RoslynImmutableInterlocked.InterlockedInitialize(ref _lazyRequiredMembers, requiredMembers);
+            }
+            else
+            {
+                // Ensure that, after this method, _lazyRequiredMembersErrorDoNotAccessDirectly is able to be accessed safely.
+                Interlocked.MemoryBarrier();
```

Comment:
`_lazyRequiredMembers` is a struct. I can't really make a "sentinel" value with it, and thus can't use it to represent what I need.

Reply:
Thanks. I was looking at the wrong `ImmutableSegmentedDictionary` (non-generic one is a class).
Check to see what Aleksey thinks, but I would rather wrap the struct and stay in a less error-prone concurrency pattern.

Reply:
Let's discuss this offline. I think we shouldn't invent new ways of handling lazy initialization. For structs we usually declare wrapper classes in order to ensure atomic read/write operations. See, for example, ```TypeWithAnnotations.Boxed```. I probably could be convinced otherwise if this situation is unique in some way.

Reply:
I'm ok with doing a `Boxed` struct or similar here. I was generally trying to avoid an allocation, but I'm fine with saving that for if we decide there's a perf issue.